### PR TITLE
Add SDPA and SDPA_decode as custom ops

### PR DESCRIPTION
**Pull request by LPanosTT**
NOTE: This PR is dependent on this uplift as it contains the logic for modelling/executing `scaled_dot_product_attention` in tt-mlir: https://github.com/tenstorrent/tt-xla/pull/1478

### Ticket
Related to: https://github.com/tenstorrent/tt-xla/issues/1365

### Problem description
We wish to directly use the ttnn scaled dot product attention ops in the vLLM plugin.

### What's changed
- Model `scaled_dot_product_attention` and `scaled_dot_product_attention_decode` as custom torch ops
- Add unit tests to test the op implementation

### Checklist
- [X] New/Existing tests provide coverage for changes

### CODE REVIEW

File: python_package/tt_torch/custom_ops.py [hunk 1]
```diff
@@ -64,6 +64,213 @@ def _(tensor: torch.Tensor, argument_type: str, name: str = None) -> torch.Tenso
...
+) -> torch.Tensor:
+    return torch.zeros_like(query)
+
+
+@torch.library.custom_op(
+    "tt::scaled_dot_product_attention_decode",
+    mutates_args=[],
+    device_types=["xla", "cpu"],
+)
+def scaled_dot_product_attention_decode(
```
Comment (nvukobratTT) [inline python_package/tt_torch/custom_ops.py:164]: Is SDPA decoding a custom op request coming from vLLM? If yes, how does it target native PyTorch?

Comment (LPanosTT) [reply python_package/tt_torch/custom_ops.py:164]: SDPA decode is a ttnn op designed specifically for running SDPA in the decode phase of the model. Added here so we can use it.

File: python_package/tt_torch/custom_ops.py [hunk 2]
```diff
@@ -64,6 +64,213 @@ def _(tensor: torch.Tensor, argument_type: str, name: str = None) -> torch.Tenso
     return tensor.clone()
 
 
+@torch.library.custom_op(
+    "tt::scaled_dot_product_attention", mutates_args=[], device_types=["xla", "cpu"]
+)
+def scaled_dot_product_attention(
```
[the remark this answers is not in the document]
Comment (LPanosTT) [reply python_package/tt_torch/custom_ops.py:70]: I suppose it could be. I just kept it in line with the other ops I've piped through directly from ttnn. I'll try it out.

[the remark this answers is not in the document]
Comment (LPanosTT) [reply python_package/tt_torch/custom_ops.py:70]: I can make it but it doesn't lower to `ttir.scaled_dot_product_attention` without the infra to do so in tt-mlir. If there turns out to be issues with custom calls in general then we should probably make all of our custom ops composite instead. What are your concerns?

File: python_package/tt_torch/custom_ops.py (hunk 1 above)
Comment (sgligorijevicTT) [reply python_package/tt_torch/custom_ops.py:164]: Torch does expose [SDPA](url) as an operation for model writers, https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html . We might want to hook into that at some point. On GPU this tries to dispatch to flashattention package, xformers, or their internal implementation of SDPA. vLLM probably targets flash attention differently.

File: python_package/tt_torch/custom_ops.py [hunk 3]
```diff
@@ -64,6 +64,213 @@ def _(tensor: torch.Tensor, argument_type: str, name: str = None) -> torch.Tenso
...
+        return stablehlo_custom_call.stablehlo_custom_call(
+            inputs,
+            "tt.scaled_dot_product_attention_decode",
+            [query.shape],
+            [query.dtype],
+            frontend_attributes=frontend_attributes,
+        )
+
+    elif query.device.type == "cpu":
+        # TODO(@LPanosTT): Model the behavior of the op when an attention_sink is provided.
```
[the remark this answers is not in the document]
Comment (sgligorijevicTT) [reply python_package/tt_torch/custom_ops.py:228]: Attention sinks are used inside gpt-oss model, basic idea is to have attention logits that don't correspond to any token pair, to allow softmax to sum to less than 1.
GPT-OSS report cites this blogpost: https://www.evanmiller.org/attention-is-off-by-one.html

File: python_package/tt_torch/custom_ops.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nvukobratTT) [reply python_package/tt_torch/custom_ops.py:70]: My concern is Shardy. If you're not using composite op with SHLO as a reference, how is Shardy handling this scenario?

File: python_package/tt_torch/custom_ops.py [hunk 4]
```diff
@@ -64,6 +64,213 @@ def _(tensor: torch.Tensor, argument_type: str, name: str = None) -> torch.Tenso
...
+        key.shape[-1] == query.shape[-1]
+    ), "key and query must have the same head size."
+
+    assert (
+        query.shape[1] % key.shape[1] == 0
+    ), "num_heads must be divisible by num_kv_heads."
+
+    # The CPU implementation of this op will funtion correctly if this invariant is not satisfied.
+    # However, this custom op is intended to exactly replicate the behavior of the ttnn op, so we will enforce this invariant.
+    assert query.shape[2] % 32 == 0, "query sequence length must be divisible by 32."
```
[the remark this answers is not in the document]
Comment (nvukobratTT) [reply python_package/tt_torch/custom_ops.py:100]: > The ProgramConfig is not currently something we model in tt-mlir.

Why? 

> Also, we are creating a torch integration. However, these ops aren't really meant for the user to use on their own, although they could if they wanted. It is for us to use in our vLLM plugin.

We're implementing these as custom operations, and they are exposed to users. That means that some of them will probably use them, especially if interested in greater performance. Also PyTorch [SPDA](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) is most used operation in torch, and I assume it'll be targeted on our end as well.

File: python_package/tt_torch/custom_ops.py (hunk 1 above)
Comment (nvukobratTT) [reply python_package/tt_torch/custom_ops.py:164]: I shared related comments in the two threads above. Also, tagging @acolicTT as he's currently enabling composite ops through torch path.

File: python_package/tt_torch/custom_ops.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (LPanosTT) [reply python_package/tt_torch/custom_ops.py:100]: >why?
No op in our ttnn dialect besides matmul models the ProgramConfig attribute. This can be done in the future for SDPA if need be. Even if we wish to do this now we'd first need to update tt-mlir and amend this branch. Which could be done separately.

>We're implementing these as custom operations, and they are exposed to users. That means that some of them will probably use them, especially if interested in greater performance. Also PyTorch [SPDA](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) is most used operation in torch, and I assume it'll be targeted on our end as well.

I'm not sure what you want me to exactly. Lower torch SDPA directly to ttir sdpa? That isn't possible since SHLO doesn't have an SDPA op. We could implement a fusion pattern in tt-mlir and probably should. But my highest priority is the vLLM plugin at the moment.

[the remark this answers is not in the document]
Comment (LPanosTT) [reply python_package/tt_torch/custom_ops.py:100]: I can wrap it in a composite op, but it will be out of place with the other custom ops as of now.

[the remark this answers is not in the document]
Comment (nvukobratTT) [reply python_package/tt_torch/custom_ops.py:100]: > I can wrap it in a composite op, but it will be out of place with the other custom ops as of now.

`mark_argument_attributes` is a good candidate to stay as a custom call. We're using it for mapping inputs, so it's not the actual op that we're encapsulating. 

Regarding fill and update cache, those are good candidates for composite ops as we're encapsulating some scatter (or combination of) ops.

File: python_package/tt_torch/custom_ops.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AleksKnezevic) [reply python_package/tt_torch/custom_ops.py:70]: shardy will have to work with custom calls since we won't be able to implement all ops as composites (i.e. sparse MoE). Figuring out how to get these through shardy is a separate task and should be done as a separate PR. This PR is blocking vLLM support for a single-chip model that a customer wants, so I don't think we should block it's progress until we resolve shardy issues.

[the remark this answers is not in the document]
Comment (AleksKnezevic) [reply python_package/tt_torch/custom_ops.py:70]: @nvukobratTT, in regards to shardy, it's not obvious to me that shard specs would propagate through composite ops the same as they would for fused kernels, nor that we would want them to. While SDPA is mathematically equivalent to MM+Softmax+MM that's not what's being done in the op. If shardy sees it as separate ops, it could allow illegal propagation (i.e. sharding on sequence length) which would then not allow us to fuse it into SDPA.
